source/system/structs/HashMap.ooc

```diff
@@ -95,262 +60,149 @@ acX31Hash: func <K> (key: K) -> SizeT {
-	else if (T size == Pointer size)
-		pointerHash
-	else if (T size == UInt size)
-		intHash
-	else
-		murmurHash
-}
-
 HashMap: class <K, V> {
-	_count: SizeT
-	capacity: SizeT
-	hashKey: Func <K> (K) -> SizeT
-	buckets: HashEntry[]
-	keys: VectorList<K>
+	_count: Int
+	_capacity: Int
+	_buckets: _HashEntry<K, V>[]
+
 	count ::= this _count
-	isEmpty ::= this keys empty
+	isEmpty ::= this _count == 0
+	capacity ::= this _capacity
 
-	init: func { init(3) }
-	init: func ~withCapacity (=capacity) {
+	init: func ~default { this init(32) }
+	init: func (=_capacity) {
 		this _count = 0
-		this buckets = HashEntry[capacity] new()
-		this keys = VectorList<K> new(32, false)
-		this hashKey = getStandardHashFunc(K)
+		this _buckets = _HashEntry[this _capacity] new()
 	}
 	free: override func {
-		for (i in 0 .. this buckets length)
-			this buckets[i] free()
-		(this buckets, this keys) free()
+		this clear()
+		this _buckets free()
 		super()
 	}
-	getEntry: func (key: K, result: HashEntry*) -> Bool {
-		hash: SizeT = this hashKey(key) % this capacity
-		entry := this buckets[hash]
-		success := false
-		if (entry key != null)
-			while (true) {
-				if (This _keyEquals(entry key as K, key)) {
-					if (result)
-						result@ = entry
-					success = true
-					break
-				}
-				if (entry next)
-					entry = entry next@
-				else
-					break
-			}
-		success
+	put: func (key: K, value: V) -> V {
+		if ((this _count as Float / this _capacity as Float) > 0.75f)
+			this resize(this _count * 2)
+		existingValue: V = this remove(key)
+		this _count += 1
+		hashkey := This hash(key) % this capacity
+		entry := _HashEntry<K, V> new(key, value)
+		entry _next = this _buckets[hashkey]
+		this _buckets[hashkey] = entry
+		existingValue
+	}
+	get: func ~nulldefault (key: K) -> V {
+		this get(key, null)
+	}
+	get: func (key: K, defaultValue: V) -> V {
+		result: V = defaultValue
+		if (this contains(key)) {
+			hashkey := This hash(key) % this capacity
+			entry := this _buckets[hashkey]
+			while (!this _keyEquals(key, entry _key))
+				entry = entry _next
+			result = entry _value
+		}
+		result
 	}
-	getEntryForHash: func (key: K, hash: SizeT, result: HashEntry*) -> Bool {
-		entry := this buckets[hash]
-		success := false
-		if (entry key != null)
-			while (true) {
-				if (This _keyEquals(entry key as K, key)) {
-					if (result)
-						result@ = entry
-					success = true
-					break
-				}
-				if (entry next)
-					entry = entry next@
-				else
-					break
+	clear: func {
+		for (index in 0 .. this capacity)
+			if (this _buckets[index] != null) {
+				this _buckets[index] free()
+				this _buckets[index] = null
 			}
-		success
-	}
-	copy: func -> This<K, V> {
-		copy := This<K, V> new()
-		f := func (k: K, v: V) { copy put(k, v) }
-		this each(f)
-		(f as Closure) free()
-		copy
-	}
-	merge: func (other: This<K, V>) -> This<K, V> {
-		c := this copy()
-		c merge~inplace(other)
-		c
-	}
-	merge: func ~inplace (other: This<K, V>) -> This<K, V> {
```

Comment:
My implementation also lacks `copy` and `merge` functions, we can add those later (including the hassle of getting them to work properly) later if we need them.